### PR TITLE
ASoC: SOF: ipc4: Fetch CPC from tplg mod_cfg or from manifest

### PR DESCRIPTION
**Pull request by ujfalusi**
Hi,

the MOD_INIT_INSTANCE message always contains a base_cfg struct and in there the CPC value must be set to non 0 to indicate the processing need of the instantiated module with ti's configuration.
Currently we have one module  specific token to pass a single, static CPC value in topology. Some module has this set, some does not, but in any case it is not reflecting the real processing need based on the configuration, parameters.

The firmware's manifest however contains mod_cfg array which contains CPC value for the module's different configuration, the CPC number should be looked up by matching the obs/ibs in manifest mod_cfg and in the runtime base_cfg obs/ibs and pick the exact matched CPC or maximum CPC.

Since the manifest is part of the firmware binary, it is hard to modify and might contain out of date or non correct value, hence we should have a way to provide similar mod_cfg type of info via tplg and use that if it is present.

This series implements this CPC lookup and configuration for the modules.

### CODE REVIEW

File: sound/soc/sof/ipc4-loader.c [hunk 1]
```diff
@@ -423,6 +423,62 @@ int sof_ipc4_reload_fw_libraries(struct snd_sof_dev *sdev)
...
+	u32 cpc_pick = 0;
+	u32 max_cpc = 0;
+	int i;
+
+	/* The CPC is set from topology, ignoring manifest */
+	if (basecfg->cpc)
+		return;
+
+	if (!fw_module || !fw_module->fm_config)
+		return;
```
Comment (ranj063) [inline sound/soc/sof/ipc4-loader.c:448]: what happens in this case? Why don't we use the ibs*obs/1000 in this case?

File: sound/soc/sof/ipc4-loader.c [hunk 2]
```diff
@@ -112,6 +112,7 @@ static ssize_t sof_ipc4_fw_parse_ext_man(struct snd_sof_dev *sdev,
 				return -EINVAL;
 			}
 
+			fw_module->fm_config = &fm_config[fm_entry->cfg_offset];
```
Comment (plbossart) [inline sound/soc/sof/ipc4-loader.c]: so for all cases of saving a pointer to reuse later, you'll see me ask:

How safe if this, and is the pointer valid always?

File: sound/soc/sof/ipc4-priv.h [hunk 3]
```diff
@@ -28,11 +28,13 @@ enum sof_ipc4_mtrace_type {
 /**
  * struct sof_ipc4_fw_module - IPC4 module info
  * @sof_man4_module: Module info
+ * @fm_config: Pointer to the module config start of the module
  * @m_ida: Module instance identifier
  * @private: Module private data
  */
 struct sof_ipc4_fw_module {
 	struct sof_man4_module man4_module_entry;
+	const struct sof_man4_module_config *fm_config;
```
Comment (plbossart) [inline sound/soc/sof/ipc4-priv.h]: what does 'fm' stand for?

Should this be config_start instead?

File: sound/soc/sof/ipc4-loader.c [hunk 4]
```diff
@@ -423,6 +423,62 @@ int sof_ipc4_reload_fw_libraries(struct snd_sof_dev *sdev)
...
 }
 
+/**
+ * sof_ipc4_update_basecfg_from_manifest - Update the base config from manifest
+ * @sdev: SOF device
+ * @fw_module: pointer struct sof_ipc4_fw_module to parse
+ * @basecfg: Pointer to the base_config to update
+ *
+ * Note: currently only used to update the CPC value if it is not set from
+ * topology already.
```
Comment (plbossart) [inline sound/soc/sof/ipc4-loader.c]: but why would we need to update the base config? You're not describing why we would need to modify a base config?

File: sound/soc/sof/ipc4-loader.c [hunk 5]
```diff
@@ -423,6 +423,62 @@ int sof_ipc4_reload_fw_libraries(struct snd_sof_dev *sdev)
...
+		dev_dbg(sdev->dev,
+			"%s: No direct CPC match based on obs/ibs %u/%u\n",
+			fw_module->man4_module_entry.name, basecfg->obs, basecfg->ibs);
+		cpc_pick = max_cpc;
+	}
+
+	if (cpc_pick) {
+		dev_dbg(sdev->dev, "%s: CPC from manifest: %u\n",
+			fw_module->man4_module_entry.name, cpc_pick);
+	} else {
```
Comment (plbossart) [inline sound/soc/sof/ipc4-loader.c]: can this branch be reached?

Is there a case where max_cpc == 0?

File: sound/soc/sof/ipc4-loader.c [hunk 6]
```diff
@@ -423,6 +423,62 @@ int sof_ipc4_reload_fw_libraries(struct snd_sof_dev *sdev)
...
+		cpc_pick = basecfg->ibs * basecfg->obs / 1000;
```
Comment (plbossart) [inline sound/soc/sof/ipc4-loader.c]: I am lost on why this formula would help? 

````
struct sof_ipc4_base_module_cfg {
	uint32_t cpc; /* the max count of Cycles Per Chunk processing */
	uint32_t ibs; /* input Buffer Size (in bytes)  */
	uint32_t obs; /* output Buffer Size (in bytes) */
	uint32_t is_pages; /* number of physical pages used */
	struct sof_ipc4_audio_format audio_fmt;
} __packed __aligned(4);
````

How would multiplying buffer sizes give us a valid number of cycles? That's super odd.

**Review by plbossart — CHANGES_REQUESTED** [review]
Didn't get the problem statement nor the suggestion solution.

File: sound/soc/sof/ipc4-loader.c (hunk 1 above)
Comment (ujfalusi) [reply sound/soc/sof/ipc4-loader.c:448]: With the new version we are going to use that at the end.

File: sound/soc/sof/ipc4-loader.c (hunk 2 above)
Comment (ujfalusi) [reply sound/soc/sof/ipc4-loader.c]: @plbossart, yes, I have asked this and since we are keeping the firmware loaded all the time, this is OK.

File: sound/soc/sof/ipc4-priv.h (hunk 3 above)
Comment (ujfalusi) [reply sound/soc/sof/ipc4-priv.h]: renamed it to `fw_mod_cfg`, as it is coming from the firmware's manifest, bundled together.

File: sound/soc/sof/ipc4-loader.c (hunk 6 above)
Comment (ujfalusi) [reply sound/soc/sof/ipc4-loader.c]: I can leave this out and just send 0 to the firmware (as we do in some cases atm), but that is not good for the firmware.
If the module is missing this information then there is really no way that we can guess as each module can have different requirements. Based on the number I have seen around, this formula gives a number in the ballpark of measured numbers among different ones.
In general bigger the buffer, more processing is needed, but I cannot figure out better way atm.

File: sound/soc/sof/ipc4-loader.c (hunk 5 above)
Comment (ujfalusi) [reply sound/soc/sof/ipc4-loader.c]: Currently in tgl the manifest only have 0 as CPC among all the configurations, so yes, it can be hit.
In the new version I have changed this as well.

File: sound/soc/sof/ipc4-loader.c (hunk 4 above)
Comment (ujfalusi) [reply sound/soc/sof/ipc4-loader.c]: New version is limited to basecfg->cpc updating, so the Note is dropped.

**Comment by ujfalusi** [comment]
> Didn't get the problem statement nor the suggestion solution.

Updated the cover letter with relevant background information.

**Comment by ujfalusi** [comment]
Changes since v1:
- limit the manifest parsing to CPC lookup to clarify the functionality
- Add support for tplg  mod_cfg parsing

Related sof/topology2 PR: https://github.com/thesofproject/sof/pull/7348

File: sound/soc/sof/ipc4-topology.c [hunk 7]
```diff
@@ -1108,6 +1108,14 @@ static int sof_ipc4_init_audio_fmt(struct snd_sof_dev *sdev,
 	if (available_fmt->num_output_formats && i < available_fmt->num_output_formats)
 		base_config->obs = available_fmt->output_pin_fmts[i].buffer_size;
 
+	/* Update the base configuration from the module manifest */
+	sof_ipc4_update_cpc_from_manifest(sdev, swidget->module_info, base_config);
+	if (!base_config->cpc) {
+		base_config->cpc = base_config->obs * base_config->ibs / 1000;
+		dev_dbg(sdev->dev, "Using calculated (obs*ibs/1000) CPC for %s: %u\n",
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-topology.c]: looking at the numbers again, it is likely that obs*ibs/100 gives the correct approximation.

File: sound/soc/sof/ipc4-loader.c (hunk 6 above)
Comment (plbossart) [reply sound/soc/sof/ipc4-loader.c]: CPC is cycle per chunk. How is the 'chunk' related to the input and output buffers?

Undocumented assumptions left and right with all this. @ujfalusi @ranj063 @kv2019i @lgirdwood unless we are able to explain in layman terms what the cpc is and how it's determine I am not hot on submitting those patches upstream. When it looks like a hack, etc.

Comment (ujfalusi) [reply sound/soc/sof/ipc4-loader.c]: @plbossart, I tend to agree, so let's leave it as 0 if it is not provided and think of another fallback path as a followup, which should be kernel only I'm afraid.

**Comment by ujfalusi** [comment]
Changes since v2:
- Drop the "fallback" calculation of CPC , leave it 0 if it cannot be acquired from any of the sources
- Rename one missed struct name (ipc4_audio_fmt_num_tokens -> ipc4_module_params_num_tokens)
- Print the final module ibs/obs/cpc numbers unconditionally
- Use consistent order in printing: ibs, obs, cpc
- The tplg mod_cfg array reading is fixed (thanks to @ranj063 and @jsarha)

**Review by jsarha — COMMENTED** [review]
Hmm, shouldn't the logic go so that if we do find an exact ibs & obs match from topology, then the manifest entry does not override it any more? Or have I misunderstood how this should go? At least the commit description in "ASoC: SOF: ipc4-loader/topology: Update CPC value from manifest if missing" suggest otherwise.

In other words, which setting has the priority, if the ibs&obs match is found from the both in the manifest and in the topology?

**Comment by ujfalusi** [comment]
> Hmm, shouldn't the logic go so that if we do find an exact ibs & obs match from topology, then the manifest entry does not override it any more? Or have I misunderstood how this should go? At least the commit description in "ASoC: SOF: ipc4-loader/topology: Update CPC value from manifest if missing" suggest otherwise.

This is the logic.
 
> In other words, which setting has the priority, if the ibs&obs match is found from the both in the manifest and in the topology?

the priority of the CPC is:
CPC is set in topology for the module as a single override value
CPC can be found in tplg mod_cfg with exact match based on obs/ibs
CPC can be found in manifest mod_cfg with exact match based on obs/ibs
Maximum CPC value for the module from topology mod_cfg
Maximum CPC value for the module from manifest mod_cfg

File: sound/soc/sof/ipc4-loader.c [hunk 8]
```diff
@@ -428,10 +428,12 @@ int sof_ipc4_reload_fw_libraries(struct snd_sof_dev *sdev)
  * @sdev: SOF device
  * @fw_module: pointer struct sof_ipc4_fw_module to parse
  * @basecfg: Pointer to the base_config to update
+ * @exact_match: Only consider exact matches based on obs/ibs
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-loader.c:431]: Stale line about the removed CPC fallback calculation, needs to be removed.

File: sound/soc/sof/ipc4-loader.c [hunk 9]
```diff
@@ -423,6 +423,53 @@ int sof_ipc4_reload_fw_libraries(struct snd_sof_dev *sdev)
...
+	if (!cpc_pick) {
+		dev_dbg(sdev->dev,
+			"%s: No exact CPC match from manifest (obs/ibs: %u/%u)\n",
+			fw_module->man4_module_entry.name, basecfg->obs, basecfg->ibs);
+
+		cpc_pick = max_cpc;
+	}
+
+	dev_dbg(sdev->dev, "%s: CPC from manifest: %u\n",
+		fw_module->man4_module_entry.name, cpc_pick);
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-loader.c:473]: Stale line about the removed CPC fallback calculation, needs to be removed.

**Comment by ujfalusi** [comment]
@jsarha, the check is within the functions to return if the basecfg->cpc is not 0, so they will not override it. I wanted to avoid excess amount of `if (base_config->cpc) goto out;` at every step in ipc4-topology.c

**Comment by greg-intel** [comment]
SOFCI TEST

**Review by jsarha — DISMISSED** [review]
About the problem statement and the solution, there may be gaps to fill, but code-wise I could not find anything wrong.

File: sound/soc/sof/ipc4-loader.c (hunk 2 above)
Comment (plbossart) [reply sound/soc/sof/ipc4-loader.c]: ok

**Comment by ujfalusi** [comment]
Changes since v3:
- Use step-by step cpc validity check fro better documentation in sof_ipc4_init_base_module_cfg() as suggested by @jsarha
- Add final fallback to platform's maximum CPS number (iow, maximum core frequency), which is stored per platform family
- If this is not available either, assume that the DSP can run up to 400MHz (all supported Intel DSPs are running at this speed).

File: sound/soc/sof/ipc4-topology.c [hunk 10]
```diff
@@ -1114,9 +1205,33 @@ sof_ipc4_init_base_module_cfg(struct snd_sof_dev *sdev,
...
+	 * Initialize the module's cpc with the stored global value from
+	 * topology
+	 */
+	base_config->cpc = module_params->cpc;
+	if (base_config->cpc)
+		goto out;
 
-	/* Update base_config->cpc from the module manifest */
-	sof_ipc4_update_cpc_from_manifest(sdev, swidget->module_info, base_config);
+	/* Update base_config->cpc from the module manifest - direct match */
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-topology.c]: comment needs update for the two `sof_ipc4_update_cpc_from_tplg()` call.

**Comment by ujfalusi** [comment]
Changes since v4:
- Fix the comment around sof_ipc4_update_cpc_from_tplg()

File: sound/soc/sof/ipc4-loader.c [hunk 11]
```diff
@@ -112,16 +112,12 @@ static ssize_t sof_ipc4_fw_parse_ext_man(struct snd_sof_dev *sdev,
 				return -EINVAL;
 			}
 
-			/* a module's config is always the same size */
-			fw_module->bss_size = fm_config[fm_entry->cfg_offset].is_bytes;
 
 			dev_dbg(sdev->dev,
 				"module %s: UUID %pUL cfg_count: %u, bss_size: %#x\n",
 				fm_entry->name, &fm_entry->uuid, fm_entry->cfg_count,
-				fw_module->bss_size);
```
Comment (RanderWang) [inline sound/soc/sof/ipc4-loader.c:121]: it should be used to calculate module memory usage and much like CPC case that we have two sources of bss_size, one in module extended manifest and one in topology. Currently we only set a unknown number 280 to bss_size in topology for copier only and not set it for other modules. This function follows ref driver and should use 
fw_module->bss_size, not base_cfg->is_page or we should check base_cfg->is_page like CPC case as
````c
       if (base_cfg->is_page)
                 bss = base_cfg->is_page;
      else
                bss = fw_module->bss_size;
````

````c
static void
sof_ipc4_update_pipeline_mem_usage(struct snd_sof_dev *sdev, struct snd_sof_widget *swidget,
				   struct sof_ipc4_base_module_cfg *base_config)
{
	struct sof_ipc4_fw_module *fw_module = swidget->module_info;
	struct snd_sof_widget *pipe_widget;
	struct sof_ipc4_pipeline *pipeline;
	int task_mem, queue_mem;
	int ibs, bss, total;

	ibs = base_config->ibs;
	bss = base_config->is_pages; // should use fw_module->bss_size
````

Comment (ujfalusi) [reply sound/soc/sof/ipc4-loader.c:121]: @RanderWang, OK, I drop this patch from the series.
Handling of this 'is_pages' is not a subject of this series and I'm anyways uncertain of what the variables are:
`fm_config[fm_entry->cfg_offset].is_bytes` is in bytes, so
`fw_module->bss_size` is also in bytes
`base_config->is_pages` is in ????
`bss` likely should be in bytes or in ???? ? Probably bytes.
thus
`base_config->is_pages` should be in bytes?

Comment (RanderWang) [reply sound/soc/sof/ipc4-loader.c:121]:
````c
struct sof_ipc4_base_module_cfg {
	uint32_t cpc; /* the max count of Cycles Per Chunk processing */
	uint32_t ibs; /* input Buffer Size (in bytes)  */
	uint32_t obs; /* output Buffer Size (in bytes) */
	uint32_t is_pages; /* number of physical pages used */
	struct sof_ipc4_audio_format audio_fmt;
} __packed __aligned(4);
````

Comment (ujfalusi) [reply sound/soc/sof/ipc4-loader.c:121]: @RanderWang, thinking about this: we can still drop bss_size as we save the pointer to fw_mod_cfg and it is available over there.

Comment (RanderWang) [reply sound/soc/sof/ipc4-loader.c:121]: yes, agree

File: sound/soc/sof/ipc4-priv.h [hunk 12]
```diff
@@ -69,6 +69,7 @@ struct sof_ipc4_fw_library {
  * @max_num_pipelines: max number of pipelines
  * @max_libs_count: Maximum number of libraries support by the FW including the
  *		    base firmware
+ * @max_kcps: maximum cycles per second * 1000, iow the maximum core frequency
```
Comment (RanderWang) [inline sound/soc/sof/ipc4-priv.h:72]: why we need to introduce max_cpc ?  The cpc should be set in manifest or we should return error to notify developer to fix this issue

Comment (ujfalusi) [reply sound/soc/sof/ipc4-priv.h:72]: this is max_**kcps**, not cpc, this is not set in manifest, it is the maximum clock speed the DSP can run if we want to put it in human readable terms: **K**ilo **C**ycles **P**er **S**edond.
CPC is is cycles per chunk and a chunk is 1 sec (at least I have only seen this convention).

**Review by RanderWang — APPROVED** [review]
LGTM

**Comment by ujfalusi** [comment]
SOFCI TEST

File: sound/soc/sof/ipc4-pcm.c [hunk 13]
```diff
@@ -465,9 +465,9 @@ static int sof_ipc4_pcm_dai_link_fixup_rate(struct snd_sof_dev *sdev,
 					    struct snd_pcm_hw_params *params,
 					    struct sof_ipc4_copier *ipc4_copier)
 {
-	struct sof_ipc4_pin_format *pin_fmts = ipc4_copier->available_fmt.input_pin_fmts;
```
Comment (ranj063) [inline sound/soc/sof/ipc4-pcm.c:468]: typo in commit "renmae"

File: sound/soc/sof/ipc4-loader.c [hunk 14]
```diff
@@ -423,6 +423,53 @@ int sof_ipc4_reload_fw_libraries(struct snd_sof_dev *sdev)
 	return ret;
 }
 
+/**
```
Comment (ranj063) [inline sound/soc/sof/ipc4-loader.c:426]: where are we doing the "use the ibs*obs/1000 formula" in this patch?

**Comment by ujfalusi** [comment]
Closing, replaced by #4315
